frontend/src/metabase/visualizations/echarts/cartesian/model/index.ts

```diff
@@ -86,10 +87,19 @@ export const getCartesianChartModel = (
   const seriesModels = getSortedSeriesModels(unsortedSeriesModels, settings);
 
   const seriesDataKeys = seriesModels.map(seriesModel => seriesModel.dataKey);
-  const dataset =
-    rawSeries[0].card.display === "scatter"
-      ? getScatterPlotDataset(rawSeries, cardsColumns)
-      : getJoinedCardsDataset(rawSeries, cardsColumns);
+
+  let dataset;
+  switch (rawSeries[0].card.display) {
+    case "scatter":
+      dataset = getScatterPlotDataset(rawSeries, cardsColumns);
+      break;
+    case "waterfall":
+      dataset = getWaterfallDataset(rawSeries[0].data.rows, cardsColumns[0]);
```

Comment:
In a later PR I made this use a separate property `waterfallDataset`, to preserve the existing `dataset` for use in dynamic behaviors.

https://github.com/metabase/metabase/pull/36925

I didn't end up updating the previous PRs to avoid having to change and rebase too many things.